File: leetcode054.py
class Solution:
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        res = []
        if len(matrix)==0:
            return res
        length = len(matrix)*len(matrix[0])
        i = 0
        j = 0
        m = 0
        n = 0
        res.append(matrix[m][n])
        matrix[m][n] = None
        i = i + 1
        while i < length:
            if j%4 == 0:

                while n+1 < len(matrix[0]) and matrix[m][n+1] !=None :
                    n = n+1
                    res.append(matrix[m][n])
                    matrix[m][n] = None
                    i=i+1
                j=j+1
            if j%4 == 1:

                while m+1 < len(matrix) and matrix[m+1][n] !=None:
                    m = m + 1
                    res.append(matrix[m][n])
                    matrix[m][n] = None
                    i=i+1
                j=j+1
            if j % 4 == 2:

                while n - 1 >=0 and matrix[m][n -1] != None:
                    n = n - 1
                    res.append(matrix[m][n])
                    matrix[m][n] = None
                    i = i + 1
                j = j + 1
            if j % 4 == 3:

                while m - 1 >=0 and matrix[m - 1][n] != None :
                    m = m - 1
                    res.append(matrix[m][n])
                    matrix[m][n] = None
                    i = i + 1
                j = j + 1
        return res

if __name__ == "__main__":
    s = Solution()
    print(s.spiralOrder([[ 1, 2, 3 ],[ 4, 5, 6 ],[ 7, 8, 9 ]]))

File: leetcodec++/leetcode054.cpp
//
// Created by 黄敏辉 on 2019-07-15.
//

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<vector<int>> dirs{{0,1}, {1,0}, {0, -1}, {-1, 0}};
        vector<int> res;
        int l1 = matrix.size();
        if (l1 == 0){
            return res;
        }
        int l2 = matrix[0].size();
        if (l2 == 0){
            return res;
        }

        vector<int> Nsteps{l2, l1-1};

        int idir = 0;
        int ir = 0, ic = -1;
        while (Nsteps[idir % 2]){
            for (int i = 0; i < Nsteps[idir % 2]; i++){
                ir += dirs[idir][0];
                ic += dirs[idir][1];
                res.push_back(matrix[ir][ic]);
            }
            Nsteps[idir%2]--;
            idir = (idir + 1) % 4;

        }
        return res;
    }
};